### DISCUSSION TITLE
Problem with saving the data on the sd card

### DISCUSSION THREAD

**Discussion by FugaziScript** [pschatzmann/arduino-audio-tools#389] [Q&A] [answered]
I'm using esp32 with the pcm5102a DAC and PCM1808 ADC. I can make a loopback from ADC to DAC as well as show the data from ADC on serial plotter, but I have a problem saving the file on the sd card. The file on the card has a name but it's size is 0B. Is it a known issue? Can I save files with 32bit depth or am I doing something wrong?

My code is similar to [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-sd-audiokit/streams-audiokit-sd-audiokit.ino)

AudioFile = SD.open("/new.raw", FILE_WRITE);
copier.begin(AudioFile, i2sStream); 
...
AudioFile = SD.close(); (After a few seconds copier.copy() in void loop())

**Comment by pschatzmann** [comment, marked as answer]
The data is only saved when you close the file with AudioFile.close()!

**Comment by mgesteiro** [comment]
hello @FugaziScript,

> I'm using esp32 with the pcm5102a DAC and PCM1808 ADC. I can make a loopback from ADC to DAC as well as show the data from ADC on serial plotter

I'm currently trying to make a PCM1808 ADC module work myself, but I'm having problems setting it up: as far as I've read, an external clock is required (and I don't have one).

Is it really necessary external circuitry or can the clock be just provided by the ESP32 itself?

I'll appreciate any information you could share for a working schematic including the PCM1808 ADC and an ESP32.

Kind regards!

**Reply by Doekle** [reply]
you can use the clock of an ESP, use the master clock to be set in the configuration. (in the pins struct) you need to set the multiple as well perhaps (depends on the da converter)


i2s_pin_config_t i2s_mic_pins = {
  .bck_io_num = I2S_MIC_SERIAL_CLOCK,
  .ws_io_num = I2S_MIC_LEFT_RIGHT_CLOCK,
  .data_out_num = I2S_PIN_NO_CHANGE,
  .data_in_num = I2S_MIC_SERIAL_DATA
//  .mck_io_num = 0, // master clock pin, not used, need also i2s_mclk_multiple_t I2S_MCLK_MULTIPLE_384  at init to work.
};

**Reply by FugaziScript** [reply]
You don't need the external clock for the pcm1808. The esp32 can provide it via pin_mck. The pcm1808 returns the data as 32 bits_per_sample and I2SConfig needs to be configured that way. If you want to save the data using any encoder you have to use NumberFormatConverterStreamT<int16_t, int32_t>(*RecordingStream). For the loopback it's the same. As for the pinout the FMY, MD1, MD0 are all grounded and the I2S pinout is self explanatory.
